File: src/render.ts
/**
 * Output formatting and rendering for workspace status.
 */

import type { FileChange, RepoSnapshot, SummaryState, VerboseRow } from './types.js';
import { badgeForChange, fileChangesFromSnapshot } from './changes.js';
import {
  extractTicketId,
  formatBranchWithMerge,
  formatCheckoutRepoLabel,
  isAttentionSyncNote,
  sortedUnique,
  splitEntries,
  sanitizePath,
  trimVal,
  visibleWidth,
} from './helpers.js';

export interface WorkspaceStatusRenderInput {
  snapshots: RepoSnapshot[];
  summary: SummaryState;
  verbose: {
    cleanDefault: VerboseRow[];
    cleanNonDefault: VerboseRow[];
    changeRepos: VerboseRow[];
    repoWidth: number;
    branchWidth: number;
  };
  showVerbose: boolean;
}

export function formatFileChanges(filesStr: string): string {
  const entries = splitEntries(filesStr);
  let added = 0,
    modified = 0,
    deleted = 0;
  for (const e of entries) {
    const status = e[0];
    if (status === 'A') added++;
    else if (status === 'M') modified++;
    else if (status === 'D') deleted++;
    else if (status === 'R') modified++;
  }
  const parts: string[] = [];
  if (added > 0) parts.push(`➕ ${added} added`);
  if (modified > 0) parts.push(`📝 ${modified} modified`);
  if (deleted > 0) parts.push(`➖ ${deleted} deleted`);
  return parts.length > 0 ? `(${parts.join(', ')})` : '';
}

export function formatUntrackedChanges(filesStr: string): string {
  const entries = splitEntries(filesStr);
  if (entries.length === 0) return '';
  return `(➕ ${entries.length} added)`;
}

interface FileTreeNode {
  dirs: Map<string, FileTreeNode>;
  files: FileChange[];
}

function padVisible(value: string, width: number): string {
  return value + ' '.repeat(Math.max(0, width - visibleWidth(value)));
}

function fileStatusEmoji(status: string): string {
  const m: Record<string, string> = { A: '➕', M: '📝', D: '➖', R: '🔄' };
  return m[status] ?? '';
}

function extractFilePath(entry: string): string {
  const tab = entry.indexOf('\t');
  if (tab < 0) return '';
  const status = entry[0];
  const rest = entry.slice(tab + 1);
  if (status === 'R') {
    const secondTab = rest.indexOf('\t');
    return secondTab >= 0 ? trimVal(rest.slice(secondTab + 1)) : trimVal(rest);
  }
  return trimVal(rest);
}

export function renderVerbose(
  rows: VerboseRow[],
  repoWidth: number,
  branchWidth: number,
): string[] {
  const syncWidth = Math.max('Sync'.length, ...rows.map((r) => visibleWidth(r.sync)));
  const filesWidth = Math.max('Files'.length, ...rows.map((r) => visibleWidth(r.files)));
  const lines = [
    `${padVisible('Repo', repoWidth)}  ${padVisible('Branch', branchWidth)}  ${padVisible('Sync', syncWidth)}  ${padVisible('Files', filesWidth)}`,
  ];
  for (const r of rows) {
    lines.push(
      `${padVisible(r.repo, repoWidth)}  ${padVisible(r.branch, branchWidth)}  ${padVisible(r.sync, syncWidth)}  ${padVisible(r.files, filesWidth)}${r.note ? `  ${r.note}` : ''}`,
    );
  }
  return lines;
}

export function renderTrackedFileEntries(filesStr: string): string[] {
  const entries = splitEntries(filesStr);
  const lines: string[] = [];
  for (const e of entries) {
    const filepath = sanitizePath(extractFilePath(e));
    if (!filepath) continue;
    const emoji = fileStatusEmoji(e[0]);
    if (emoji) lines.push(`      ${emoji} ${filepath}`);
  }
  return lines;
}

export function renderUntrackedFileEntries(filesStr: string): string[] {
  const entries = splitEntries(filesStr);
  return entries
    .map((e) => sanitizePath(trimVal(e)))
    .filter(Boolean)
    .map((p) => `      ➕ ${p}`);
}

function repoLabel(snapshot: RepoSnapshot): string {
  const repo = formatCheckoutRepoLabel(snapshot);
  const ticketId = extractTicketId(snapshot.branch);
  return ticketId ? `${repo} (${ticketId})` : repo;
}

function fileDisplay(change: FileChange): string {
  const name = change.path.split('/').pop() ?? change.path;
  if (change.oldPath) {
    const oldName = change.oldPath.split('/').pop() ?? change.oldPath;
    return `${badgeForChange(change)} ${oldName} -> ${name}`;
  }
  return `${badgeForChange(change)} ${name}`;
}

function addTreeChange(root: FileTreeNode, change: FileChange): void {
  const parts = change.path.split('/').filter(Boolean);
  const fileName = parts.pop();
  if (!fileName) return;
  let node = root;
  for (const dir of parts) {
    let child = node.dirs.get(dir);
    if (!child) {
      child = { dirs: new Map(), files: [] };
      node.dirs.set(dir, child);
    }
    node = child;
  }
  node.files.push(change);
}

function collapseNode(name: string, node: FileTreeNode): { name: string; node: FileTreeNode } {
  let collapsedName = name;
  let collapsedNode = node;
  while (collapsedNode.files.length === 0 && collapsedNode.dirs.size === 1) {
    const [[childName, childNode]] = collapsedNode.dirs;
    collapsedName += `/${childName}`;
    collapsedNode = childNode;
  }
  return { name: collapsedName, node: collapsedNode };
}

function renderTreeNode(node: FileTreeNode, prefix: string): string[] {
  const dirEntries = [...node.dirs.entries()]
    .map(([name, child]) => collapseNode(name, child))
    .sort((a, b) => a.name.localeCompare(b.name));
  const fileEntries = [...node.files].sort((a, b) => a.path.localeCompare(b.path));
  const items: Array<{ label: string; child?: FileTreeNode }> = [
    ...dirEntries.map((entry) => ({ label: entry.name, child: entry.node })),
    ...fileEntries.map((change) => ({ label: fileDisplay(change) })),
  ];

  const lines: string[] = [];
  items.forEach((item, index) => {
    const last = index === items.length - 1;
    lines.push(`${prefix}${last ? '└─' : '├─'} ${item.label}`);
    if (item.child) lines.push(...renderTreeNode(item.child, `${prefix}${last ? '   ' : '│  '}`));
  });
  return lines;
}

function syncLabel(snapshot: RepoSnapshot): string {
  const repo = formatCheckoutRepoLabel(snapshot);
  const ticketId = extractTicketId(snapshot.branch);
  let label = `${repo} [${snapshot.branch}]`;
  if (ticketId) label += ` (${ticketId})`;
  if (snapshot.syncNote) label += ` - ${snapshot.syncNote}`;
  return label;
}

function renderRepoChangeLines(snapshot: RepoSnapshot): string[] {
  const root: FileTreeNode = { dirs: new Map(), files: [] };
  for (const change of fileChangesFromSnapshot(snapshot)) addTreeChange(root, change);
  return [`  📦 ${repoLabel(snapshot)}`, ...renderTreeNode(root, '     ')];
}

function appendSectionGap(lines: string[]): void {
  if (lines.length > 0) lines.push('');
}

function appendSyncGroupLines(
  lines: string[],
  label: string,
  repos: string[],
  snapshotMap: Map<string, RepoSnapshot>,
): void {
  if (repos.length === 0) return;
  lines.push(label);
  for (const repo of repos) {
    const snapshot = snapshotMap.get(repo);
    if (snapshot) lines.push(`    - ${syncLabel(snapshot)}`);
  }
}

function branchSummaryLabel(snapshot: RepoSnapshot): string {
  const repo = formatCheckoutRepoLabel(snapshot);
  const ticketId = extractTicketId(snapshot.branch);
  const base = ticketId ? `${repo} (${ticketId})` : `${repo} [${snapshot.branch}]`;
  return formatBranchWithMerge(base, snapshot.mergedIntoDefault);
}

function appendBranchGroupLines(
  lines: string[],
  label: string,
  repos: string[],
  snapshotMap: Map<string, RepoSnapshot>,
): void {
  if (repos.length === 0) return;
  lines.push(label);
  for (const repo of repos) {
    const snapshot = snapshotMap.get(repo);
    if (snapshot) lines.push(`    - ${branchSummaryLabel(snapshot)}`);
  }
}

function linkedWorktreeSummaryLabel(snapshot: RepoSnapshot): string {
  return branchSummaryLabel(snapshot);
}

function appendLinkedWorktreesSection(
  lines: string[],
  linkedPaths: string[],
  snapshotMap: Map<string, RepoSnapshot>,
): void {
  if (linkedPaths.length === 0) return;
  appendSectionGap(lines);
  lines.push(`🔗 Linked worktrees (${linkedPaths.length}):`);
  for (const repo of linkedPaths) {
    const snapshot = snapshotMap.get(repo);
    if (snapshot) lines.push(`    - ${linkedWorktreeSummaryLabel(snapshot)}`);
  }
}

export function renderWorkspaceStatus(input: WorkspaceStatusRenderInput): string[] {
  const { snapshots, summary, verbose, showVerbose } = input;
  const lines: string[] = [];
  const snapshotMap = new Map(snapshots.map((s) => [s.repo, s]));
  const linkedPaths = sortedUnique([...summary.linkedWorktrees]);

  if (showVerbose) {
    const rows = [...verbose.cleanDefault, ...verbose.cleanNonDefault, ...verbose.changeRepos];
    lines.push(...renderVerbose(rows, verbose.repoWidth, verbose.branchWidth));
  }

  if (snapshots.length === 0) {
    appendSectionGap(lines);
    lines.push('ℹ️ No git repos found');
    return lines;
  }

  const totalChanges =
    summary.changesUncommitted.size +
    summary.changesStaged.size +
    summary.changesBoth.size +
    summary.changesUntracked.size;
  const totalSync = summary.syncBehind.size + summary.syncAhead.size + summary.syncDiverged.size;
  const totalBranches =
    summary.branchFeature.size +
    summary.branchBugfix.size +
    summary.branchChore.size +
    summary.branchRelease.size +
    summary.branchUnknown.size;
  const attentionRepos = snapshots
    .filter((s) => isAttentionSyncNote(s.syncNote))
    .sort((a, b) => a.repo.localeCompare(b.repo));

  if (
    totalChanges === 0 &&
    totalSync === 0 &&
    totalBranches === 0 &&
    attentionRepos.length === 0
  ) {
    appendSectionGap(lines);
    lines.push('✅ All repos clean and up-to-date');
    appendLinkedWorktreesSection(lines, linkedPaths, snapshotMap);
    return lines;
  }

  if (totalChanges > 0) {
    const reposWithChanges = sortedUnique([
      ...summary.changesUncommitted,
      ...summary.changesStaged,
      ...summary.changesBoth,
      ...summary.changesUntracked,
    ]);
    appendSectionGap(lines);
    lines.push('File changes');
    reposWithChanges.forEach((repo, index) => {
      const snapshot = snapshotMap.get(repo);
      if (snapshot && index > 0) lines.push('');
      if (snapshot) lines.push(...renderRepoChangeLines(snapshot));
    });
  }

  if (totalSync > 0) {
    appendSectionGap(lines);
    lines.push(`🔄 Sync status (${totalSync}):`);
    appendSyncGroupLines(lines, '  ⬇️ behind:', sortedUnique([...summary.syncBehind]), snapshotMap);
    appendSyncGroupLines(lines, '  ⬆️ ahead:', sortedUnique([...summary.syncAhead]), snapshotMap);
    appendSyncGroupLines(
      lines,
      '  🔀 diverged:',
      sortedUnique([...summary.syncDiverged]),
      snapshotMap,
    );
  }

  if (totalBranches > 0) {
    appendSectionGap(lines);
    lines.push(`🌿 Branches (${totalBranches}):`);
    appendBranchGroupLines(
      lines,
      '  🚧 feature:',
      sortedUnique([...summary.branchFeature]),
      snapshotMap,
    );
    appendBranchGroupLines(
      lines,
      '  🐛 bugfix:',
      sortedUnique([...summary.branchBugfix]),
      snapshotMap,
    );
    appendBranchGroupLines(
      lines,
      '  🔧 chore:',
      sortedUnique([...summary.branchChore]),
      snapshotMap,
    );
    appendBranchGroupLines(
      lines,
      '  🚀 release:',
      sortedUnique([...summary.branchRelease]),
      snapshotMap,
    );
    appendBranchGroupLines(
      lines,
      '  🌿 unknown:',
      sortedUnique([...summary.branchUnknown]),
      snapshotMap,
    );
  }

  if (attentionRepos.length > 0) {
    appendSectionGap(lines);
    lines.push(`⚠️ Attention (${attentionRepos.length}):`);
    for (const snapshot of attentionRepos) {
      lines.push(`    - ${syncLabel(snapshot)}`);
    }
  }

  appendLinkedWorktreesSection(lines, linkedPaths, snapshotMap);

  return lines;
}

File: crates/workspace-status/src/render.rs
//! `--plain` renderer.

use std::collections::{BTreeMap, HashMap};

use crate::helpers::{
    extract_ticket_id, format_branch_with_merge, format_checkout_repo_label, is_attention_sync_note,
    sorted_unique, visible_width,
};
use crate::snapshot::{FileChange, RepoSnapshot, SummaryState, VerboseRow};

fn pad_visible(value: &str, width: usize) -> String {
    let pad = width.saturating_sub(visible_width(value));
    format!("{value}{}", " ".repeat(pad))
}

fn badge_for_change(change: &FileChange) -> &'static str {
    if change.unstaged_status.as_deref() == Some("U") || change.staged_status.as_deref() == Some("U")
    {
        return "⚠️U";
    }
    if change.staged_status.is_some() && change.unstaged_status.is_some() {
        return "🟠MS";
    }
    let status = change
        .unstaged_status
        .as_deref()
        .or(change.staged_status.as_deref());
    match status {
        Some("R") => "🟣R",
        Some("D") => "🔴D",
        Some("A") => "🟢A",
        Some(_) if change.staged_status.is_some() => "🔵S",
        _ if change.untracked => "🟢A",
        _ => "🟡M",
    }
}

fn file_display(change: &FileChange) -> String {
    let name = change.path.rsplit('/').next().unwrap_or(&change.path);
    if let Some(old) = &change.old_path {
        let old_name = old.rsplit('/').next().unwrap_or(old);
        format!("{} {old_name} -> {name}", badge_for_change(change))
    } else {
        format!("{} {name}", badge_for_change(change))
    }
}

#[derive(Default)]
struct FileTreeNode {
    dirs: BTreeMap<String, FileTreeNode>,
    files: Vec<FileChange>,
}

fn add_tree_change(root: &mut FileTreeNode, change: &FileChange) {
    let mut parts: Vec<&str> = change.path.split('/').filter(|s| !s.is_empty()).collect();
    let Some(file_name) = parts.pop() else {
        return;
    };
    let _ = file_name;
    let mut node = root;
    for dir in parts {
        node = node.dirs.entry(dir.to_string()).or_default();
    }
    node.files.push(change.clone());
}

fn collapse_node<'a>(name: String, node: &'a FileTreeNode) -> (String, &'a FileTreeNode) {
    let mut collapsed_name = name;
    let mut collapsed_node = node;
    while collapsed_node.files.is_empty() && collapsed_node.dirs.len() == 1 {
        let (child_name, child_node) = collapsed_node.dirs.iter().next().unwrap();
        collapsed_name = format!("{collapsed_name}/{child_name}");
        collapsed_node = child_node;
    }
    (collapsed_name, collapsed_node)
}

fn render_tree_node(node: &FileTreeNode, prefix: &str) -> Vec<String> {
    let mut dir_entries: Vec<(String, &FileTreeNode)> = node
        .dirs
        .iter()
        .map(|(name, child)| collapse_node(name.clone(), child))
        .collect();
    dir_entries.sort_by(|a, b| a.0.cmp(&b.0));
    let mut file_entries = node.files.clone();
    file_entries.sort_by(|a, b| a.path.cmp(&b.path));

    let mut items: Vec<(String, Option<&FileTreeNode>)> = dir_entries
        .into_iter()
        .map(|(n, c)| (n, Some(c)))
        .collect();
    items.extend(file_entries.iter().map(|c| (file_display(c), None)));

    let mut lines = Vec::new();
    let last_i = items.len().saturating_sub(1);
    for (index, (label, child)) in items.iter().enumerate() {
        let last = index == last_i;
        lines.push(format!("{prefix}{} {label}", if last { "└─" } else { "├─" }));
        if let Some(child) = child {
            let next = format!("{prefix}{}", if last { "   " } else { "│  " });
            lines.extend(render_tree_node(child, &next));
        }
    }
    lines
}

fn repo_label(snapshot: &RepoSnapshot) -> String {
    let repo = format_checkout_repo_label(snapshot);
    if let Some(ticket) = extract_ticket_id(&snapshot.branch) {
        format!("{repo} ({ticket})")
    } else {
        repo
    }
}

fn sync_label(snapshot: &RepoSnapshot) -> String {
    let repo = format_checkout_repo_label(snapshot);
    let ticket = extract_ticket_id(&snapshot.branch);
    let mut label = format!("{repo} [{}]", snapshot.branch);
    if let Some(ticket) = ticket {
        label.push_str(&format!(" ({ticket})"));
    }
    if !snapshot.sync_note.is_empty() {
        label.push_str(&format!(" - {}", snapshot.sync_note));
    }
    label
}

fn branch_summary_label(snapshot: &RepoSnapshot) -> String {
    let repo = format_checkout_repo_label(snapshot);
    let ticket = extract_ticket_id(&snapshot.branch);
    let base = if let Some(ticket) = ticket {
        format!("{repo} ({ticket})")
    } else {
        format!("{repo} [{}]", snapshot.branch)
    };
    format_branch_with_merge(&base, snapshot.merged_into_default)
}

fn render_repo_change_lines(snapshot: &RepoSnapshot) -> Vec<String> {
    let mut root = FileTreeNode::default();
    for change in &snapshot.changes {
        add_tree_change(&mut root, change);
    }
    let mut lines = vec![format!("  📦 {}", repo_label(snapshot))];
    lines.extend(render_tree_node(&root, "     "));
    lines
}

fn render_verbose(rows: &[VerboseRow], repo_width: usize, branch_width: usize) -> Vec<String> {
    let sync_width = rows
        .iter()
        .map(|r| visible_width(&r.sync))
        .max()
        .unwrap_or(0)
        .max(visible_width("Sync"));
    let files_width = rows
        .iter()
        .map(|r| visible_width(&r.files))
        .max()
        .unwrap_or(0)
        .max(visible_width("Files"));
    let mut lines = vec![format!(
        "{}  {}  {}  {}",
        pad_visible("Repo", repo_width),
        pad_visible("Branch", branch_width),
        pad_visible("Sync", sync_width),
        pad_visible("Files", files_width)
    )];
    for r in rows {
        let mut line = format!(
            "{}  {}  {}  {}",
            pad_visible(&r.repo, repo_width),
            pad_visible(&r.branch, branch_width),
            pad_visible(&r.sync, sync_width),
            pad_visible(&r.files, files_width)
        );
        if !r.note.is_empty() {
            line.push_str(&format!("  {}", r.note));
        }
        lines.push(line);
    }
    lines
}

fn append_section_gap(lines: &mut Vec<String>) {
    if !lines.is_empty() {
        lines.push(String::new());
    }
}

fn append_sync_group(
    lines: &mut Vec<String>,
    label: &str,
    repos: &[String],
    map: &HashMap<String, RepoSnapshot>,
) {
    if repos.is_empty() {
        return;
    }
    lines.push(label.to_string());
    for repo in repos {
        if let Some(snapshot) = map.get(repo) {
            lines.push(format!("    - {}", sync_label(snapshot)));
        }
    }
}

fn append_branch_group(
    lines: &mut Vec<String>,
    label: &str,
    repos: &[String],
    map: &HashMap<String, RepoSnapshot>,
) {
    if repos.is_empty() {
        return;
    }
    lines.push(label.to_string());
    for repo in repos {
        if let Some(snapshot) = map.get(repo) {
            lines.push(format!("    - {}", branch_summary_label(snapshot)));
        }
    }
}

fn append_linked_section(
    lines: &mut Vec<String>,
    linked: &[String],
    map: &HashMap<String, RepoSnapshot>,
) {
    if linked.is_empty() {
        return;
    }
    append_section_gap(lines);
    lines.push(format!("🔗 Linked worktrees ({}):", linked.len()));
    for repo in linked {
        if let Some(snapshot) = map.get(repo) {
            lines.push(format!("    - {}", branch_summary_label(snapshot)));
        }
    }
}

pub fn render_workspace_status(
    snapshots: &[RepoSnapshot],
    summary: &SummaryState,
    verbose: &(Vec<VerboseRow>, Vec<VerboseRow>, Vec<VerboseRow>, usize, usize),
    show_verbose: bool,
) -> Vec<String> {
    let mut lines = Vec::new();
    let map: HashMap<String, RepoSnapshot> = snapshots
        .iter()
        .map(|s| (s.repo.clone(), s.clone()))
        .collect();
    let linked = sorted_unique(summary.linked_worktrees.iter().cloned());

    if show_verbose {
        let mut rows = Vec::new();
        rows.extend(verbose.0.clone());
        rows.extend(verbose.1.clone());
        rows.extend(verbose.2.clone());
        lines.extend(render_verbose(&rows, verbose.3, verbose.4));
    }

    if snapshots.is_empty() {
        append_section_gap(&mut lines);
        lines.push("ℹ️ No git repos found".to_string());
        return lines;
    }

    let total_changes = summary.changes_uncommitted.len()
        + summary.changes_staged.len()
        + summary.changes_both.len()
        + summary.changes_untracked.len();
    let total_sync =
        summary.sync_behind.len() + summary.sync_ahead.len() + summary.sync_diverged.len();
    let total_branches = summary.branch_feature.len()
        + summary.branch_bugfix.len()
        + summary.branch_chore.len()
        + summary.branch_release.len()
        + summary.branch_unknown.len();
    let mut attention: Vec<&RepoSnapshot> = snapshots
        .iter()
        .filter(|s| is_attention_sync_note(&s.sync_note))
        .collect();
    attention.sort_by(|a, b| a.repo.cmp(&b.repo));

    if total_changes == 0 && total_sync == 0 && total_branches == 0 && attention.is_empty() {
        append_section_gap(&mut lines);
        lines.push("✅ All repos clean and up-to-date".to_string());
        append_linked_section(&mut lines, &linked, &map);
        return lines;
    }

    if total_changes > 0 {
        let repos = sorted_unique(
            summary
                .changes_uncommitted
                .iter()
                .chain(&summary.changes_staged)
                .chain(&summary.changes_both)
                .chain(&summary.changes_untracked)
                .cloned(),
        );
        append_section_gap(&mut lines);
        lines.push("File changes".to_string());
        for (index, repo) in repos.iter().enumerate() {
            if let Some(snapshot) = map.get(repo) {
                if index > 0 {
                    lines.push(String::new());
                }
                lines.extend(render_repo_change_lines(snapshot));
            }
        }
    }

    if total_sync > 0 {
        append_section_gap(&mut lines);
        lines.push(format!("🔄 Sync status ({total_sync}):"));
        append_sync_group(
            &mut lines,
            "  ⬇️ behind:",
            &sorted_unique(summary.sync_behind.iter().cloned()),
            &map,
        );
        append_sync_group(
            &mut lines,
            "  ⬆️ ahead:",
            &sorted_unique(summary.sync_ahead.iter().cloned()),
            &map,
        );
        append_sync_group(
            &mut lines,
            "  🔀 diverged:",
            &sorted_unique(summary.sync_diverged.iter().cloned()),
            &map,
        );
    }

    if total_branches > 0 {
        append_section_gap(&mut lines);
        lines.push(format!("🌿 Branches ({total_branches}):"));
        append_branch_group(
            &mut lines,
            "  🚧 feature:",
            &sorted_unique(summary.branch_feature.iter().cloned()),
            &map,
        );
        append_branch_group(
            &mut lines,
            "  🐛 bugfix:",
            &sorted_unique(summary.branch_bugfix.iter().cloned()),
            &map,
        );
        append_branch_group(
            &mut lines,
            "  🔧 chore:",
            &sorted_unique(summary.branch_chore.iter().cloned()),
            &map,
        );
        append_branch_group(
            &mut lines,
            "  🚀 release:",
            &sorted_unique(summary.branch_release.iter().cloned()),
            &map,
        );
        append_branch_group(
            &mut lines,
            "  🌿 unknown:",
            &sorted_unique(summary.branch_unknown.iter().cloned()),
            &map,
        );
    }

    if !attention.is_empty() {
        append_section_gap(&mut lines);
        lines.push(format!("⚠️ Attention ({}):", attention.len()));
        for snapshot in attention {
            lines.push(format!("    - {}", sync_label(snapshot)));
        }
    }

    append_linked_section(&mut lines, &linked, &map);
    lines
}
